File: weeks/week-01-foundations/src/tool-contract.ts
export type ToolPermission = "read" | "write" | "network" | "deploy";

export type ToolDescriptor = Readonly<{
  name: string;
  sideEffect: "none" | "consequential";
  permissions: readonly ToolPermission[];
  requiresApproval: boolean;
  validateInput: (input: unknown) => boolean;
  outputSchema: string;
}>;

export type ToolProposal = Readonly<{
  tool: string;
  permissions: readonly string[];
  input: unknown;
}>;

export type ValidatedToolCall = Readonly<{
  tool: string;
  permissions: readonly ToolPermission[];
  input: unknown;
  sideEffect: ToolDescriptor["sideEffect"];
}>;

const PERMISSIONS = new Set<ToolPermission>([
  "read",
  "write",
  "network",
  "deploy",
]);

function isPermission(value: string): value is ToolPermission {
  // 문자열을 닫힌 권한 집합으로 좁히는 TypeScript type guard입니다.
  return PERMISSIONS.has(value as ToolPermission);
}

export function createToolRegistry(
  descriptors: readonly ToolDescriptor[],
): ReadonlyMap<string, ToolDescriptor> {
  // descriptor 오류는 첫 모델 호출 전에 발견해야 실행 중 정책이 흔들리지 않습니다.
  const registry = new Map<string, ToolDescriptor>();
  for (const descriptor of descriptors) {
    const name = descriptor.name.trim();
    if (!/^[a-z][a-z0-9_]{2,63}$/.test(name)) {
      throw new TypeError(`도구 이름이 유효하지 않습니다: ${descriptor.name}`);
    }
    // 같은 이름의 구현이 둘이면 실제 실행 대상을 예측할 수 없습니다.
    if (registry.has(name)) throw new Error(`중복 도구 이름: ${name}`);
    // 부작용 분류와 승인 정책의 모순을 구성 단계에서 거부합니다.
    if (
      descriptor.permissions.length === 0 ||
      descriptor.permissions.some((permission) => !isPermission(permission))
    ) {
      throw new Error(`알 수 없거나 비어 있는 도구 권한: ${name}`);
    }
    if (
      new Set(descriptor.permissions).size !== descriptor.permissions.length
    ) {
      throw new Error(`중복 도구 권한: ${name}`);
    }
    if (descriptor.outputSchema.trim().length < 3) {
      throw new Error(`출력 schema가 필요합니다: ${name}`);
    }
    if (
      descriptor.sideEffect === "consequential" &&
      !descriptor.requiresApproval
    ) {
      throw new Error(`부작용 도구에는 승인이 필요합니다: ${name}`);
    }
    registry.set(
      name,
      Object.freeze({
        ...descriptor,
        name,
        permissions: Object.freeze([...descriptor.permissions]),
        outputSchema: descriptor.outputSchema.trim(),
      }),
    );
  }
  return registry;
}

export function validateToolProposal(
  registry: ReadonlyMap<string, ToolDescriptor>,
  proposal: ToolProposal,
  approvalGranted = false,
): ValidatedToolCall {
  // allowlist에 없는 도구는 구현 탐색이나 실행 전에 즉시 거부합니다.
  const descriptor = registry.get(proposal.tool);
  if (!descriptor) throw new Error(`등록되지 않은 도구: ${proposal.tool}`);
  // 요청 권한은 descriptor 권한의 부분집합이어야 하며 자동 승격하지 않습니다.
  if (
    proposal.permissions.length === 0 ||
    proposal.permissions.some((permission) => !isPermission(permission))
  ) {
    throw new Error("알 수 없거나 비어 있는 권한 요청입니다.");
  }
  const requestedPermissions = proposal.permissions.map(
    (permission) => permission as ToolPermission,
  );
  if (
    requestedPermissions.some(
      (permission) => !descriptor.permissions.includes(permission),
    )
  ) {
    throw new Error("descriptor 범위를 벗어난 권한 요청입니다.");
  }
  // 도구 이름이 맞아도 입력 계약이 다르면 executor에 전달하지 않습니다.
  if (!descriptor.validateInput(proposal.input)) {
    throw new TypeError("도구 입력 schema 검증에 실패했습니다.");
  }
  if (descriptor.requiresApproval && !approvalGranted) {
    throw new Error("부작용 도구 실행에는 사람 승인이 필요합니다.");
  }
  // 검증 결과를 동결해 executor로 이동하는 동안 권한이나 입력 참조가 바뀌지 않게 합니다.
  return Object.freeze({
    tool: descriptor.name,
    permissions: Object.freeze(requestedPermissions),
    input: proposal.input,
    sideEffect: descriptor.sideEffect,
  });
}

File: python-labs/src/agent_harness_labs/week1/tool_contract.py
"""Validate model-proposed tool calls before an executor can see them."""

from collections.abc import Callable, Iterable, Mapping
from dataclasses import dataclass
from enum import Enum
from types import MappingProxyType
from typing import TypeAlias

from agent_harness_labs.errors import ContractError

InputValidator: TypeAlias = Callable[[object], bool]


class ToolPermission(str, Enum):
    READ = "read"
    WRITE = "write"
    NETWORK = "network"
    DEPLOY = "deploy"


class SideEffect(str, Enum):
    NONE = "none"
    CONSEQUENTIAL = "consequential"


@dataclass(frozen=True)
class ToolDescriptor:
    name: str
    side_effect: SideEffect
    permissions: tuple[ToolPermission, ...]
    requires_approval: bool
    validate_input: InputValidator
    output_schema: str


@dataclass(frozen=True)
class ToolProposal:
    tool: str
    permissions: tuple[str, ...]
    input: object


@dataclass(frozen=True)
class ValidatedToolCall:
    tool: str
    permissions: tuple[ToolPermission, ...]
    input: object
    side_effect: SideEffect


def object_has_nonempty_string(key: str) -> InputValidator:
    def validate(value: object) -> bool:
        return (
            isinstance(value, Mapping)
            and isinstance(value.get(key), str)
            and bool(str(value[key]).strip())
        )

    return validate


def create_tool_registry(
    descriptors: Iterable[ToolDescriptor],
) -> Mapping[str, ToolDescriptor]:
    registry: dict[str, ToolDescriptor] = {}
    for descriptor in descriptors:
        name = descriptor.name.strip()
        if not name.isidentifier() or not name.isascii() or len(name) < 3:
            raise ContractError("INVALID_TOOL_DESCRIPTOR", "도구 이름이 유효하지 않습니다.")
        if name in registry:
            raise ContractError("DUPLICATE_TOOL", "도구 이름이 중복됩니다.")
        if not descriptor.permissions or len(set(descriptor.permissions)) != len(
            descriptor.permissions
        ):
            raise ContractError("INVALID_PERMISSION_SET", "권한 집합이 비었거나 중복됩니다.")
        if not descriptor.output_schema.strip():
            raise ContractError("INVALID_OUTPUT_SCHEMA", "출력 schema가 필요합니다.")
        if (
            descriptor.side_effect is SideEffect.CONSEQUENTIAL
            and not descriptor.requires_approval
        ):
            raise ContractError("APPROVAL_POLICY_INVALID", "부작용 도구에는 승인이 필요합니다.")
        registry[name] = descriptor
    return MappingProxyType(registry)


def validate_tool_proposal(
    registry: Mapping[str, ToolDescriptor],
    proposal: ToolProposal,
    *,
    approval_granted: bool = False,
) -> ValidatedToolCall:
    descriptor = registry.get(proposal.tool)
    if descriptor is None:
        raise ContractError("UNKNOWN_TOOL", "등록되지 않은 도구입니다.")
    try:
        permissions = tuple(ToolPermission(item) for item in proposal.permissions)
    except ValueError as error:
        raise ContractError("PERMISSION_INVALID", "알 수 없는 권한입니다.") from error
    if not permissions:
        raise ContractError("PERMISSION_INVALID", "권한 요청이 비었습니다.")
    if any(item not in descriptor.permissions for item in permissions):
        raise ContractError("PERMISSION_ESCALATION", "등록 범위를 벗어난 권한입니다.")
    if not descriptor.validate_input(proposal.input):
        raise ContractError("INPUT_SCHEMA_INVALID", "도구 입력 schema 검증에 실패했습니다.")
    if descriptor.requires_approval and not approval_granted:
        raise ContractError("APPROVAL_REQUIRED", "사람 승인이 필요합니다.")
    return ValidatedToolCall(
        tool=descriptor.name,
        permissions=permissions,
        input=proposal.input,
        side_effect=descriptor.side_effect,
    )
